Make MediaKeysStorage directory name private

szewai opened the pull request:
#### aa8f2cebaa0975801430b2421afc68884d0b3911
<pre>
Make MediaKeysStorage directory name private
<a href="https://bugs.webkit.org/show_bug.cgi?id=259985">https://bugs.webkit.org/show_bug.cgi?id=259985</a>
rdar://99984441

Reviewed by Brent Fulgham.

MediaKeysStorage uses domain string as directory name, and that could reveal information about what sites are visisted.
This patch makes MediaKeysStorage start to use hashed origin string for directory name instead.

API test: WKWebsiteDataStore.FetchAndDeleteMediaKeysData

* Source/WebCore/Modules/encryptedmedia/CDM.cpp:
(WebCore::CDM::storageDirectory const):
* Source/WebCore/Modules/encryptedmedia/MediaKeySession.cpp:
(WebCore::MediaKeySession::mediaKeysStorageDirectory const):
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeySession.cpp:
(WebCore::WebKitMediaKeySession::mediaKeysStorageDirectory const):
* Source/WebCore/Modules/storage/DummyStorageProvider.h:
* Source/WebCore/Modules/storage/StorageProvider.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::mediaKeysStorageDirectory):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerMediaKeysStorageDirectory const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::ensureMediaKeysStorageDirectoryForOrigin):
(WebCore::Page::setMediaKeysStorageDirectory):
* Source/WebCore/page/Page.h:
* Source/WebCore/storage/StorageUtilities.cpp:
(WebCore::StorageUtilities::encodeSecurityOriginForFileName):
* Source/WebCore/storage/StorageUtilities.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::computeMediaKeyFile):
(WebKit::WebsiteDataStore::migrateMediaKeysStorageIfNecessary):
(WebKit::WebsiteDataStore::resolveDirectoriesIfNecessary):
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
(WebKit::WebsiteDataStore::mediaKeysStorageOrigins):
(WebKit::WebsiteDataStore::removeMediaKeysStorage):
(WebKit::WebsiteDataStore::mediaKeysStorageSalt const):
(WebKit::WebsiteDataStore::mediaKeyOrigins): Deleted.
(WebKit::WebsiteDataStore::removeMediaKeys): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebStorageProvider.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setWebsiteDataStoreParameters):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::mediaKeysStorageDirectory const):
(WebKit::WebProcess::mediaKeysStorageSalt const):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _preferencesChanged:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:

Canonical link: <a href="https://commits.webkit.org/268737@main">https://commits.webkit.org/268737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f54fb4d34d594aebf510de442a1387c4db3152b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20146 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20581 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21204 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18801 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20377 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23820 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20743 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20267 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20366 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20268 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17505 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22896 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17438 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18305 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24578 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18515 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18481 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22548 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19074 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16190 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18275 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4929 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22616 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18899 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -342,6 +348,42 @@ void WebsiteDataStore::unregisterProcess(WebProcessProxy& process)
+
+    auto originDirectoryNames = FileSystem::listDirectory(directory);
+    // Migrate existing data to new version directory.
+    for (auto name : originDirectoryNames) {
```
I don't see how we avoid doing this work every time. How does it detect that it has already migrated the media keys to the new format?

szewai replied:
It just tries to move old directories to new version directory every time. If there is no old origin directory, then no move is needed.

brentfulgham commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -342,6 +348,42 @@ void WebsiteDataStore::unregisterProcess(WebProcessProxy& process)
+    // Migrate existing data to new version directory.
+    for (auto name : originDirectoryNames) {
+        if (name == versionName)
+            continue;
```
Here we just move to the next item in the directory. Shouldn't we `break` here if we find the new file indicating it was migrated?

szewai replied:
Existence of new version directory does not indicate successful migration of all data, e.g. process crashes during the migration and only some origin directories are moved to new version directory.

brentfulgham reviewed (approved):
Please keep an eye on the performance bots, in case the new file operations impact perf! r=me

szewai committed:
Make MediaKeysStorage directory name private
https://bugs.webkit.org/show_bug.cgi?id=259985
rdar://99984441

Reviewed by Brent Fulgham.

MediaKeysStorage uses domain string as directory name, and that could reveal information about what sites are visisted.
This patch makes MediaKeysStorage start to use hashed origin string for directory name instead.

API test: WKWebsiteDataStore.FetchAndDeleteMediaKeysData

* Source/WebCore/Modules/encryptedmedia/CDM.cpp:
(WebCore::CDM::storageDirectory const):
* Source/WebCore/Modules/encryptedmedia/MediaKeySession.cpp:
(WebCore::MediaKeySession::mediaKeysStorageDirectory const):
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeySession.cpp:
(WebCore::WebKitMediaKeySession::mediaKeysStorageDirectory const):
* Source/WebCore/Modules/storage/DummyStorageProvider.h:
* Source/WebCore/Modules/storage/StorageProvider.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::mediaKeysStorageDirectory):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerMediaKeysStorageDirectory const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::ensureMediaKeysStorageDirectoryForOrigin):
(WebCore::Page::setMediaKeysStorageDirectory):
* Source/WebCore/page/Page.h:
* Source/WebCore/storage/StorageUtilities.cpp:
(WebCore::StorageUtilities::encodeSecurityOriginForFileName):
* Source/WebCore/storage/StorageUtilities.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::computeMediaKeyFile):
(WebKit::WebsiteDataStore::migrateMediaKeysStorageIfNecessary):
(WebKit::WebsiteDataStore::resolveDirectoriesIfNecessary):
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
(WebKit::WebsiteDataStore::mediaKeysStorageOrigins):
(WebKit::WebsiteDataStore::removeMediaKeysStorage):
(WebKit::WebsiteDataStore::mediaKeysStorageSalt const):
(WebKit::WebsiteDataStore::mediaKeyOrigins): Deleted.
(WebKit::WebsiteDataStore::removeMediaKeys): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebStorageProvider.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setWebsiteDataStoreParameters):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::mediaKeysStorageDirectory const):
(WebKit::WebProcess::mediaKeysStorageSalt const):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _preferencesChanged:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:

Canonical link: https://commits.webkit.org/268737@main
